src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -0,0 +1,337 @@
+// Copyright 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "command_line_parser.h"
+#include "common.h"
+#include "doctest.h"
+#include "mock_client_backend.h"
+#include "mock_model_parser.h"
+#include "request_rate_manager.h"
+
+namespace cb = triton::perfanalyzer::clientbackend;
+
+namespace triton { namespace perfanalyzer {
+
+/// Class to test the RequestRateManager
+///
+class TestRequestRateManager {
+  public:
+    TestRequestRateManager() {
+      // Must reset this global flag every unit test. 
+      // It gets set to true when we deconstruct any load manager 
+      // (aka every unit test)
+      //
+      early_exit = false;
+
+      stats_ = std::make_shared<cb::MockClientStats>();
+    }
+
+    /// Test that the correct Infer function is called in the backend
+    ///
+    void TestInferType(bool is_async, bool is_streaming) {
+      PerfAnalyzerParameters params;
+      params.async = is_async;
+      params.streaming = is_streaming;
+
+      double request_rate = 50;
+      auto sleep_time = std::chrono::milliseconds(100);
+
+      std::unique_ptr<LoadManager> manager = CreateManager(params);
+      dynamic_cast<RequestRateManager*>(manager.get())->ChangeRequestRate(request_rate);
+      std::this_thread::sleep_for(sleep_time);
+
+      CheckInferType(params);
+    }
+
+    /// Test that the inference distribution is as expected
+    ///
+    void TestDistribution(PerfAnalyzerParameters params) {
```

Comment:
I would prefer we don't have unit tests that do explicit waiting. I have two proposal alternatives:

1. Do this kind of waiting in a CI test where wait times are already expected.
2. Mock out the calculation of the wait time, or something similar, so that when the sleep call actually gets called, it's given a 0 value, but you could still record the wait times and calculate the distribution of "calls" from them to make sure averages and latencies are correct.

Reply:
In-line with another thread, it's fine if we wanna leave it as is an later change this so that unit test explicit waiting is removed. I kinda like the mocking of the wait time idea.

Reply:
What is the concern about waiting? How long the tests take?
Long term I hope we can get rid of these, but for now to give us confidence in what we are doing this is our best option.  
A CI test would not have caught the bug I found.  
I can't mock out the wait time of the live code without significantly refactoring it, which is the one thing we can't do in this step of the process.

You can easily exclude these tests on local unit-test runs via ` -tce=*request_rate*`